setup/src/setup-haskell.ts

```diff
@@ -33,21 +35,47 @@ export default async function run(
         exec('stack', ['setup', opts.ghc.resolved])
       );
 
-    if (opts.cabal.enable)
-      await core.group('Setting up cabal', async () => {
-        await exec('cabal', ['user-config', 'update'], {silent: true});
-        const configFile = await cabalConfig();
-
-        if (process.platform === 'win32') {
-          fs.appendFileSync(configFile, 'store-dir: C:\\sr\n');
-          core.setOutput('cabal-store', 'C:\\sr');
-        } else {
-          core.setOutput('cabal-store', `${process.env.HOME}/.cabal/store`);
-        }
-
-        await exec('cabal user-config update');
-        if (!opts.stack.enable) await exec('cabal update');
-      });
+    if (opts.cabal.enable) core.info('Loading cache...');
+    await exec('cabal', ['freeze'], {silent: true});
+    const freezeHash = await hashElement('cabal.project.freeze'); // replace with c.hashFiles when exposed in actions/toolkit
+    const keys = [os, opts.ghc.resolved, freezeHash, process.env.GITHUB_SHA];
+    const key = keys.join('-');
+    const restoreKeys = [
+      keys.slice(0, 4).join('-') + '-',
+      keys.slice(0, 3).join('-') + '-',
+      keys.slice(0, 2).join('-') + '-'
+    ];
+    const cabalStore =
+      process.platform === 'win32'
+        ? 'C:\\sr'
+        : `${process.env.HOME}/.cabal/store`;
+    core.setOutput('cabal-store', cabalStore);
+    const paths = [cabalStore, 'dist-newstyle'];
+    const cacheHit = await c.restoreCache(paths, key, restoreKeys);
+    core.info(`...${cacheHit ? 'done' : 'not found'}`);
+
+    await core.group('Setting up cabal', async () => {
+      await exec('cabal', ['user-config', 'update'], {silent: true});
+      const configFile = await cabalConfig();
+
+      if (process.platform === 'win32')
+        fs.appendFileSync(configFile, 'store-dir: C:\\sr\n');
+
+      await exec('cabal user-config update');
+      if (!opts.stack.enable) await exec('cabal update');
+    });
+
+    if (cacheHit && cacheHit != keys[0]) {
```

Comment:
is `!=` intentional here? (Just curious since I've got strict equality burned into my brain at this point)